src/Language/PureScript/Sugar/TypeClasses/Deriving.hs

```diff
@@ -3,43 +3,71 @@
-import Language.PureScript.TypeChecker (checkNewtype)
+import           Prelude.Compat
+
+import           Control.Arrow (second)
+import           Control.Monad (replicateM)
+import           Control.Monad.Error.Class (MonadError(..))
+import           Control.Monad.Supply.Class (MonadSupply)
+import           Data.List (foldl', find, sortBy, unzip5)
+import qualified Data.Map as M
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Ord (comparing)
+import           Data.Text (Text)
+import           Language.PureScript.AST
+import           Language.PureScript.Crash
+import           Language.PureScript.Environment
+import           Language.PureScript.Errors
+import           Language.PureScript.Externs
+import           Language.PureScript.Kinds
+import           Language.PureScript.Names
+import           Language.PureScript.Types
+import           Language.PureScript.TypeChecker (checkNewtype)
+import           Language.PureScript.TypeChecker.Synonyms (replaceAllTypeSynonyms')
 import qualified Language.PureScript.Constants as C
 
+type SynonymMap = M.Map (Qualified (ProperName 'TypeName)) ([(Text, Maybe Kind)], Type)
+
 -- | Elaborates deriving instance declarations by code generation.
 deriveInstances
-  :: (MonadError MultipleErrors m, MonadSupply m)
-  => Module
+  :: forall m
+   . (MonadError MultipleErrors m, MonadSupply m)
+  => [ExternsFile]
+  -> Module
   -> m Module
-deriveInstances (Module ss coms mn ds exts) = Module ss coms mn <$> mapM (deriveInstance mn ds) ds <*> pure exts
+deriveInstances externs (Module ss coms mn ds exts) =
+    Module ss coms mn <$> mapM (deriveInstance mn synonyms ds) ds <*> pure exts
+  where
+    synonyms :: SynonymMap
+    synonyms =
+        M.fromList $ (externs >>= \ExternsFile{..} -> mapMaybe (fromExternsDecl efModuleName) efDeclarations)
+                  ++ mapMaybe fromLocalDecl ds
+      where
+        fromExternsDecl mn' (EDTypeSynonym name args ty) = Just (Qualified (Just mn') name, (args, ty))
+        fromExternsDecl _ _ = Nothing
+
+        fromLocalDecl (TypeSynonymDeclaration name args ty) = do
+          Just (Qualified (Just mn) name, (args, ty))
+        fromLocalDecl (PositionedDeclaration _ _ d) = fromLocalDecl d
+        fromLocalDecl _ = Nothing
+
+replaceAllTypeSynonymsM
+  :: MonadError MultipleErrors m
+  => SynonymMap
+  -> Type
+  -> m Type
+replaceAllTypeSynonymsM syns = either throwError pure . replaceAllTypeSynonyms' syns
 
 -- | Takes a declaration, and if the declaration is a deriving TypeInstanceDeclaration,
 -- elaborates that into an instance declaration via code generation.
 deriveInstance
   :: (MonadError MultipleErrors m, MonadSupply m)
   => ModuleName
+  -> SynonymMap
   -> [Declaration]
   -> Declaration
   -> m Declaration
-deriveInstance mn ds (TypeInstanceDeclaration nm deps className tys@[ty] DerivedInstance)
+deriveInstance mn _ ds (TypeInstanceDeclaration nm deps className tys@[ty] DerivedInstance)
```

Comment:
I think these need to do synonym substitution too.  They inspect data constructor arguments; for example, `deriveOrd` calls `objectType` to check for a record.